### PR TITLE
Add win_mssql_query module: Module for query MSSQL databases

### PR DESCRIPTION
**Pull request by dlazz**
##### SUMMARY
This new module allows user to query MSSQL Server databases
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
win_mssql_query

##### ANSIBLE VERSION

```
2.7
```

##### ADDITIONAL INFORMATION
Example: List all the databases on a SQL Server
```
tasks:
    - name: get databases
      win_mssql_query:
        query: "SELECT name FROM sys.databases" 
        server_instance: localhost
        server_instance_user: sa
        server_instance_password: "{{ sa }}"
      register: databases

    - debug:
        msg: "{{ item.name }}"
      loop: "{{ databases.output }}"
      when: databases.output != []
```
Output:
```

PLAY [win_mssql_query] *************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************
ok: [ansible]

TASK [get databases] ***************************************************************************************************************************************************************************
changed: [ansible-test] => {"changed": true, "output": [{"name": "master"}, {"name": "tempdb"}, {"name": "model"}, {"name": "msdb"}]}

TASK [debug] ***********************************************************************************************************************************************************************************
ok: [ansible] => (item={'name': 'master'}) => {
    "msg": "master"
}
ok: [ansible] => (item={'name': 'tempdb'}) => {
    "msg": "tempdb"
}
ok: [ansible] => (item={'name': 'model'}) => {
    "msg": "model"
}
ok: [ansible] => (item={'name': 'msdb'}) => {
    "msg": "msdb"
}

PLAY RECAP *************************************************************************************************************************************************************************************
ansible               : ok=3    changed=1    unreachable=0    failed=0
```

### CODE REVIEW

**Comment by dagwieers** [comment]
Beware that [the deadline](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule) for getting new features/modules accepted in Ansible v2.7 is nearing, it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

**Comment by jhawkesworth** [comment]
This would be a nice module to have, but I think you could give it a basic kind of idempotence without too much difficulty.
You could add 2 new parameters 'creates' and 'removes' which would take sql queries to run.  If the 'creates' query returns any results, then the main query will not be run (again).  In a similar way you could alternatively specify a 'removes' query which would work the opposite way: If the 'removes' query returns any results, then the main query would be run.  This would let you ensure that rows were not present in the database.

**Comment by dlazz** [comment]
can you give me some example?

File: lib/ansible/modules/windows/win_mssql_query.ps1 [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        }
+    }
+    catch {
+        $result.query = $Query
+        $result.database = $Database
+        Fail-Json -obj $result -message "Error: $($_.exception.message)"
+    }
+    if ($res) {
+        Switch ($res.GetType()|Select-Object -ExpandProperty Name) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.ps1]: You might be better off doing `if ($res -is [Array])` and handle that accordingly. Having codepaths based on the string of the type seems wrong to me.

File: lib/ansible/modules/windows/win_mssql_query.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+# file of the same name.
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_mssql_query
+version_added: '2.7'
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.py]: Bump to 2.8

File: lib/ansible/modules/windows/win_mssql_query.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    - Invokes a query on a MSSQL Server. This module isn't idempotent, the query will be always executed.
+    - In example, if you run a query that creates a database C(CREATE DATABASE TestDb), the first time it will success
+    - the second time it will fail because the database already exists.
+requirements:
+  - Powershell module SQLPS or Powershell module SqlServer
+options:
+  server_instance:
+    description:
+      - Name of the sql server instance to query.
+    default: computername
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.py]: Keep out the default, add a new entry to the description that says `- If not specified, the remote host will be used.`

File: lib/ansible/modules/windows/win_mssql_query.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  server_instance_user:
+    description:
+      - Name of a sql user with sufficient privileges.
+  server_instance_password:
+    description:
+      - Password for server_instance_user
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.py]: When referencing another option, enclose that in I(), e.g. `I(server_instance_user)`

File: lib/ansible/modules/windows/win_mssql_query.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  query_timeout:
+    description:
+      - Specifies the number of seconds before the queries time out.
+    default: 60
+  database:
+    description:
+      - database where the query has to be executed
+    default: master
+author:
+  - Daniele Lazzari
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.py]: Please put your github user id in brackets after the name. This is used so the Ansible bot can ping you on issues and pull requests for this module.

File: lib/ansible/modules/windows/win_mssql_query.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+notes:
+  - If the user that runs ansible has the appropriated permission on the db,
+  - you don't need to set C(server_instance_user) and C(server_instance_password).
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.py]: No need to put this in a new list entry, just put `you` in the newline where it is but remove `-`. You may want to add in that CredSSP, Kerb with cred delgation or become must be used for implicit auth to work.

File: lib/ansible/modules/windows/win_mssql_query.ps1 [hunk 7]
```diff
@@ -0,0 +1,72 @@
+#!powershell
+# -*- coding: utf-8 -*-
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.ps1]: This isn't needed in PowerShell.

File: lib/ansible/modules/windows/win_mssql_query.ps1 [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        [string]$Database,
+        [string]$ServerInstance,
+        [string]$ServerInstanceUser,
+        [Security.SecureString]$ServerInstancePassword,
+        [bool]$CheckMode
+    )
+    $output = @()
+    try {
+        if (!($CheckMode)) {
+            if (!($ServerInstanceUser) -or !($ServerInstancePassword)) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_mssql_query.ps1]: Consider using splatting here instead;

```
$sql_params = @{
    ServerInstance = $ServerInstance
    Database = $Database
    Query = $Query
    QueryTimeout = $QueryTimeout
}
if ($ServerInstanceUser -and $ServerInstancePassword) {
    $SQLCredential = New-Object TypeName System.Management.Automation.PSCredential ($ServerInstanceUser, $ServerInstancePassword)
    $sql_params["Credential"] = $SQLCredential
}

Invoke-SqlCmd @sql_params
```

The benefit to this is you only call `Invoke-SqlCmd` in one spot, also it's a lot easier to deal with if statements when you aren't checking the false result.

**Review by jborean93 — COMMENTED** [review]
Looks great, just a few comments but thanks for the work you've done so far.

**Comment by dlazz** [comment]
@jborean93 thank you very much for your advices! I will make all the changes you have suggestad as soon as I can!

File: lib/ansible/modules/windows/win_mssql_query.ps1 (hunk 1 above)
Comment (dlazz) [reply lib/ansible/modules/windows/win_mssql_query.ps1]: I don't like it too, but the problem here is that `if ($res -is [Array])` returns true whether is an Object[] or a DataRow.

File: lib/ansible/modules/windows/win_mssql_query.ps1 (hunk 8 above)
Comment (dlazz) [reply lib/ansible/modules/windows/win_mssql_query.ps1]: Excellent, I didn't know it was possible, I was looking for a solution like this and I'll definitely do it.

File: lib/ansible/modules/windows/win_mssql_query.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_mssql_query.ps1]: That's interesting [DataRow](https://docs.microsoft.com/en-us/dotnet/api/system.data.datarow?view=netframework-4.7.2) doesn't show any indication that it is an Array. You could also try `$res.GetType().IsArray` and iterate from there.

Should we be only getting the first entry if the value is `Object[]`? It seems like if it is an array we would be returning all `DataRows` in that array, e.g. make `$result.output` a list of `DataRow` values.

Comment (dlazz) [reply lib/ansible/modules/windows/win_mssql_query.ps1]: Sorry, I was wrong :(
```
$res = Invoke-Sqlcmd -ServerInstance mysql -Query "Select top 2 name from sys.databases"
Is Array: $($res -is [Array])
Is Array: True
$res = Invoke-Sqlcmd -ServerInstance mysql -Query "Select top 1 name from sys.databases"
Is Array: $($res -is [Array])
Is Array: False
```
So I think that the code above could become:
```
if ($res) {
    if ($res -is [Array]){
        $Columns = $res[0].Table.Columns.Caption
    } 
    else {
        $Columns = $res.Table.Columns.Caption
    }
    $output = @($res |Select-Object -Property $Columns)
}
```

**Comment by dlazz** [comment]
Closing this because I made a mess with rebase. I will open antother one

**Pull request closed without merging**